### PR TITLE
Split solana_sdk.h

### PR DESCRIPTION
**Pull request by jackcmay**
#### Problem

`solana_sdk.h` has become a kitchen sink for everything BPf program C related

#### Summary of Changes

Split the mono-header into logical units

Fixes #

### CODE REVIEW

File: programs/bpf/c/src/deprecated_loader/deprecated_loader.c [hunk 1]
```diff
@@ -2,8 +2,9 @@
  * @brief Example C-based BPF program that prints out the parameters
  * passed to it
  */
-#include <solana_sdk.h>
-#include <deserialize_deprecated.h>
+#include <sol_types.h>
+#include <sol_log.h>
+#include <sol_deserialize_deprecated.h>
```
Comment (mvines) [inline programs/bpf/c/src/deprecated_loader/deprecated_loader.c:7]: totally just a style thing, but did you consider this?
```suggestion
#include <sol/types.h>
#include <sol/log.h>
#include <sol/deserialize_deprecated.h>
```

Comment (jackcmay) [reply programs/bpf/c/src/deprecated_loader/deprecated_loader.c:7]: heh, I thought you were in the anti-directory camp

File: sdk/bpf/c/inc/solana_sdk.h [hunk 2]
```diff
@@ -1,741 +1,20 @@
...
-}
-#endif
-
-#ifdef __cplusplus
-}
-#endif
+ * @brief Solana C-based BPF program types utility functions
+ */
+
+#include <deserialize_deprecated.h>
```
Comment (mvines) [inline sdk/bpf/c/inc/solana_sdk.h:6]: Should this be `#include <sol_deserialize_deprecated.h>`?  If not the usage of `sol_deserialize_deprecated.h` elsewhere in this PR is probably wrong

Comment (mvines) [reply sdk/bpf/c/inc/solana_sdk.h:6]: oops, nm

**Review by mvines — APPROVED** [review]
much improvement!

File: programs/bpf/c/src/deprecated_loader/deprecated_loader.c (hunk 1 above)
Comment (mvines) [reply programs/bpf/c/src/deprecated_loader/deprecated_loader.c:7]: I'm in the anti-shift key camp!  So / and - always wins over _ on my keyboard.

File: sdk/bpf/c/inc/solana_sdk.h (hunk 2 above)
Comment (jackcmay) [reply sdk/bpf/c/inc/solana_sdk.h:6]: Yeah, it should be

File: programs/bpf/c/src/deprecated_loader/deprecated_loader.c (hunk 1 above)
Comment (jackcmay) [reply programs/bpf/c/src/deprecated_loader/deprecated_loader.c:7]: heh, yeah, old ways.  how about `sol-types.h`?

Comment (mvines) [reply programs/bpf/c/src/deprecated_loader/deprecated_loader.c:7]: kebab case for C headers isn't really a norm though I think, so snake feels better in this case sadly.  That's why I didn't suggest `sol-deserialize-deprecated.h`.  I can take or leave the change of `sol_` to `sol/`, totally your call

**Review by mvines — APPROVED** [review]
:bowtie:

**Pull request merged**